### PR TITLE
rectangle with border.width=2 width=0

### PR DESCRIPTION
**Pull request by igosoft**
This test will check if no border is rendered for Rectangle when it doesn't have width height or implicitWidth implicitHeght and when its border.width is greater than zero.

### CODE REVIEW

**Comment by igosoft** [comment]
Please review. Can I merge it ?

**Comment by ChALkeR** [comment]
@igosoft Why was the second rectangle changed on the image?

**Comment by igosoft** [comment]
Or why the 2nd rect had red border :)? In qml second rect doesn't have red border

**Comment by ChALkeR** [comment]
@igosoft Damn. Sorry. It looks like my mistake when merging.

The border should be there, the border-radius shouldn't. `RectangleBorder.qml` should be a copy of `RectangleBorderChildren.qml`, but without the gray border on the parent element.

Not sure how did the tests pass.

Will fix ASAP.

**Comment by ChALkeR** [comment]
Looks like `.getImageData()` returns only partial result — the top of the image is there, but everything from below point is just transparent black. Ow.

**Comment by ChALkeR** [comment]
See #118 for the fix.
Btw, I guess this is blocked by #118.

**Comment by ChALkeR** [comment]
@pavelvasev #118 landed, could you rebase this PR and redo the png, please?

**Comment by igosoft** [comment]
@ChALkeR Please review and merge. I'm getting lost with my changes that start depend to each other :)

I had to add an extra if-condition checking Rectangle width and height when border with is set to zero. In html a border can exit without a content. I also merged #114 with #108 .

**Comment by ChALkeR** [comment]
@igosoft Ah, it seems like you were merging between branches instead of rebasing, this is why the diff looks huge here.

I will rebase this.

**Comment by ChALkeR** [comment]
@igosoft Rebased.

What was the «remove» meant to do?

File: src/qtcore/qml/elements/QtQuick/Rectangle.js [hunk 1]
```diff
@@ -26,7 +26,11 @@ registerQmlType({
         this.css.borderWidth = this.border.width + 'px';
```
Comment (ChALkeR) [inline src/qtcore/qml/elements/QtQuick/Rectangle.js]: Also, it's needed here.

File: src/qtcore/qml/elements/QtQuick/Rectangle.js [hunk 2]
```diff
@@ -26,7 +26,11 @@ registerQmlType({
...
     });
     this.border.widthChanged.connect(this, function(newVal) {
-        this.css.borderWidth = newVal + 'px';
+        if (this.width > 0 && this.height > 0){
+            this.css.borderWidth = newVal + "px";
+        } else {
+            this.css.borderWidth = "0px";
+        }
```
Comment (ChALkeR) [inline src/qtcore/qml/elements/QtQuick/Rectangle.js]: This won't work in any other case than zero-sized rectangle.

**Comment by ChALkeR** [comment]
This change won't work with any other case than zero-sized rectangle.
Making the `border = Max(Max(width, height)/2, border)` would work when `Max(width, height)/2` is even, but wouldn't work when it's odd.

The suggested fix is:
```
QMLRectangle.prototype.$calculateBorder = function() {
  if (2 * this.border.width > this.width || 2 * this.border.width > this.height) {
    return '0 ' + '0 ' + this.height + 'px ' + this.width + 'px';
  }
  return this.border.width + 'px';
}
```

and use it to set the border from both those places.

**Comment by igosoft** [comment]
@ChALkeR Looks good but what if rect.width or rect.heigth is equal zero ? and there is #107 - I have some ideas how to solve it.

**Comment by ChALkeR** [comment]
This doesn't look ready yet.
Why was the testcase removed here?

**Comment by ChALkeR** [comment]
@igosoft I rebased this. Could you please explain the «remove» commit (f68aca3)?

File: src/qtcore/qml/elements/QtQuick/Rectangle.js [hunk 3]
```diff
@@ -12,31 +12,49 @@ registerQmlType({
     this.border = new QObject(this);
     createSimpleProperty("color", this.border, "color");
     createSimpleProperty("int", this.border, "width");
-
-    this.border.color = 'black';
+    this.border.color = "black";
```
Comment (ChALkeR) [inline src/qtcore/qml/elements/QtQuick/Rectangle.js]: Unrelated.

File: src/qtcore/qml/elements/QtQuick/Rectangle.js [hunk 4]
```diff
@@ -12,31 +12,49 @@ registerQmlType({
...
-    this.border.color = 'black';
+    this.border.color = "black"; 
     this.border.width = 1;
 
     this.colorChanged.connect(this, function(newVal) {
         this.css.backgroundColor = QMLColor(newVal);
     });
     this.radiusChanged.connect(this, function(newVal) {
-        this.css.borderRadius = newVal + 'px';
+        this.css.borderRadius = newVal + "px";
```
Comment (ChALkeR) [inline src/qtcore/qml/elements/QtQuick/Rectangle.js]: Unrelated.

File: tests/Render/Simple/RectangleBorder.qml [hunk 5]
```diff
@@ -1,37 +0,0 @@
...
-    border.width: 1
-  }
-  Rectangle {
-    x: 60
-    border.width: 0
-    border.color: 'red'
-    width: 10
-    height: 10
-  }
-}
```
Comment (ChALkeR) [inline tests/Render/Simple/RectangleBorder.qml]: Why was the test removed?

File: src/qtcore/qml/elements/QtQuick/Rectangle.js [hunk 6]
```diff
@@ -12,31 +12,49 @@ registerQmlType({
...
     });
 
     this.color = "white";
     this.radius = 0;
-    this.css.borderWidth ='0px';
-    this.css.borderStyle = 'solid';
-    this.css.borderColor = 'black';
+    this.css.borderWidth = "0px";
+    this.css.borderStyle = "solid";
+    this.css.boxSizing = "border-box";
```
Comment (ChALkeR) [inline src/qtcore/qml/elements/QtQuick/Rectangle.js]: Not needed. All three lines here should be kept as they were.

File: src/qtcore/qml/elements/QtQuick/Rectangle.js [hunk 7]
```diff
@@ -12,31 +12,49 @@ registerQmlType({
...
 
     this.$drawItem = function(c) {
+        //descr("draw rect", this, ["x", "y", "width", "height", "color"]);
+        //descr("draw rect.border", this.border, ["color", "width"]);
```
Comment (ChALkeR) [inline src/qtcore/qml/elements/QtQuick/Rectangle.js]: What is this?

File: tests/Render/Simple/RectangleBorder.qml (hunk 5 above)
Comment (igosoft) [inline tests/Render/Simple/RectangleBorder.qml]: @ChALkeR I'm not sure but I think I removed not the file I wanted.

**Comment by igosoft** [comment]
@ChALkeR Do you have any idea what is the reason of build failure ?

**Comment by Plaristote** [comment]
@igosoft sounds like the parser couldn't parse qmlweb's source.
You can reproduce the issue using `npm run test`. The actual details of the parse error are unfortunately unavailable for now, it's one of the shortcomings of these JS_Parse_Error.
You can probably track it down using `git bisect`.

**Comment by igosoft** [comment]
events.js:141
      throw er; // Unhandled 'error' event
      ^
Error
    at new **JS_Parse_Error** (eval at <anonymous> (/home/travis/build/qmlweb/qmlweb/node_modules/gulp-uglify/node_modules/uglify-js/tools/node.js:22:1), <anonymous>:1526:18)

We have two **JS_Parse_Error** functions
1. src/uglify/parse.js 
2. src/qtcore/qml/lib/parser.js 

can that be a root cause ?

**Comment by henrikrudstrom** [comment]
@igosoft made a pull request with a (perhaps temporary) fix to the Js_Parse_Error in #141

**Comment by igosoft** [comment]
@henrikrudstrom  still the same :(

**Comment by henrikrudstrom** [comment]
@igosoft This error came from the minification process (my fix was only for javascript in qml files), there was a syntax error in Rectangle.js fixed that, but some render tests still fail. 

PS if you run `QMLWEB_SAVE_RENDER=1 gulp test` the images rendered by phantom are saved in `tmp/`

**Comment by igosoft** [comment]
@henrikrudstrom Thanks for the review/help/fix. I can fix this one tonight :)

**Comment by ChALkeR** [comment]
Will take a look at this soon.

**Comment by ChALkeR** [comment]
Squashed and cleaned up a bit.

**Comment by ChALkeR** [comment]
This does need further fixes, but I'll merge this now, we can fix the remaining issues later.

**Comment by ChALkeR** [comment]
Thanks!
